### PR TITLE
fix(pdf): wait for SPA content before printing

### PR DESCRIPTION
**Pull request by Kikobeats**
## Problem

SPA / app-shell pages often look “settled” before real content exists:

1. **Bot-check / verification** interstitials pass early settle signals, then get serialized as a blank or checkpoint PDF.
2. **Overflow shells** keep the document height ≈ viewport while real content lives in an inner scroller — so `window` scroll never hydrates lazy sections, and `fullPage` / `page.pdf()` only see the viewport-sized shell.
3. **Print CSS** (`@media print`) can collapse the on-screen layout; callers sometimes need `mediaType: 'screen'`.

Before this PR, PDF `waitUntil: 'auto'` did not share screenshot’s readiness path, and neither PDF nor fullPage screenshot prepared overflow documents consistently.

## Before vs after

```mermaid
flowchart LR
  subgraph before [Before]
    B1[goto settle] --> B2[PDF / viewport shot]
    B2 --> B3[Blank shell or checkpoint]
  end

  subgraph after [After]
    A1[goto settle] --> A2[paintSignals + isPageReady]
    A2 -->|not ready| A3[probe / hydrate scroll]
    A3 --> A2
    A2 -->|ready| A4[prepareFullDocument]
    A4 --> A5[scroll overflow → expand → capture]
  end
```

| | Before | After |
|---|---|---|
| PDF readiness | Settle signals only; no CLI `isPageReady` | Same gate as screenshot: `paintSignals` + `isPageReady` (bot-check aware) |
| Lazy / overflow content | Scroll `window` only (or not at all) | Shared `prepareFullDocument`: wait → scroll tallest overflow → expand |
| fullPage screenshot | Often viewport-sized shell | Viewport probes for readiness, then full-document prep + fullPage shot |
| Site-specific layout | — | Stays with the caller (`--styles`, `--scripts`, `--mediaType=screen`, …) |

```mermaid
sequenceDiagram
  participant CLI
  participant PDF as pdf.prepare
  participant Ready as waitForReady / isPageReady
  participant Prep as prepareFullDocument
  participant Chrome

  CLI->>PDF: url + opts
  PDF->>Chrome: goto (mediaType via goto)
  PDF->>Ready: paintSignals + isPageReady
  alt not ready
    Ready->>Chrome: cheap JPEG probe
    opt non-white + remaining budget
      Ready->>Chrome: one hydrate scroll
    end
  end
  Ready-->>PDF: isReady
  alt isReady
    PDF->>Prep: scroll + expandOverflow
    Prep-->>PDF: documentPrepared
  end
  PDF->>Chrome: page.pdf (expand only if prepared)
```

## What changed

- **`@browserless/pdf`** — `auto` mode mirrors screenshot readiness; overflow expand at print time only for pages that cleared prepare (`WeakMap` per page).
- **`@browserless/screenshot`** — new `prepare-full-document.js` shared by fullPage screenshot and PDF; one `OVERFLOW_MIN_PX` for wait / scroll / expand.
- **CLI** — PDF gets `isPageReady`; `--timeout` also applied to browser/context (so long prepares are not killed by the 30s default).
- **Callers own layout** — no site-specific selectors in core. Example for a slide-style report:

```bash
browserless pdf 'https://example.com/report' \
  --path=output.pdf --mediaType=screen --timeout=90000 \
  --width=960 --height=679 --margin=0 --scale=1 \
  --styles='.report-page-anchor{break-after:page}'

browserless screenshot 'https://example.com/report' \
  --path=output.png --fullPage --mediaType=screen --timeout=90000
```

## Test plan

- [ ] PDF of an SPA/report with `--mediaType=screen` shows real content (not skeleton/checkpoint)
- [ ] Same URL with `--fullPage` screenshot is tall (document flow), not viewport-only
- [ ] Bot-check / verification page is not expanded/printed as the “ready” document
- [ ] Default PDF (no `mediaType`) still works for ordinary printable pages
- [ ] `packages/browserless` + `@browserless/screenshot` tests pass

---

> [!NOTE]
> **Medium Risk**
> Changes core capture timing, DOM mutation (overflow expand), and default PDF/screenshot behavior for SPAs; broad test updates but real-site edge cases remain.
> 
> **Overview**
> **PDF and screenshot now share one readiness and full-document path** so SPAs, bot-check pages, and inner scrollers behave consistently before capture.
> 
> **`@browserless/pdf`** — `waitUntil: 'auto'` uses the same flow as screenshot: `waitForReady` / `paintSignals`, optional **`isPageReady`**, JPEG probes when not ready, a single **`tryHydrateScroll`** when the probe isn’t white, then **`prepareFullDocument`** when ready. The public **`waitForDom`** option is removed; non-`auto` navigations still run **`prepareFullDocument`**. **`render`** runs **`expandOverflow`** before **`page.pdf()`**.
> 
> **`@browserless/screenshot`** — Scroll, DOM quiet, and overflow expand move into **`prepare-full-document.js`** (tallest overflow scroller, hydrate scroll, expand). **`fullPage`** waits via viewport shots + **`checkPageReady`**, then prepares and captures expanded. **`waitForDom`** is dropped; helpers (**`prepareFullDocument`**, **`checkPageReady`**, **`paintSignals`**, etc.) are re-exported for PDF.
> 
> **CLI** — The **`pdf`** command forwards CLI **`isPageReady`** (verification markers); **`--timeout`** applies to browser launch and **`createContext`**.
> 
> Docs and tests updated; a regression test asserts every helper PDF imports from screenshot is re-exported.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9cb7224f18e46de94557a5865383da059ba7a4e7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/pdf/src/index.js [hunk 1]
```diff
@@ -137,49 +173,91 @@ module.exports = ({ goto, ...gotoOpts } = {}) => {
...
       // and an unknown viewport skips the fast path rather than dropping the
       // taller-than-viewport guard. A gate that timed out never settled, so
-      // don't trust its partial snapshot: fall through to the blank check.
-      const painted = ready.painted > 0 || (ready.text >= TEXT_PAINTED_MIN && ready.fonts)
-      if (
+      // don't trust its partial signals: fall through to the blank check.
+      let isReady =
         !ready.timedOut &&
-        painted &&
+        isPaintedContent(ready) &&
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/pdf/src/index.js:115]: Fixed in fa8f709c: paint fast path now runs `isPageReady` (with goto `response` + page meta) before skipping the screenshot poll.

File: packages/pdf/src/index.js [hunk 2]
```diff
@@ -137,49 +164,83 @@ module.exports = ({ goto, ...gotoOpts } = {}) => {
...
+          const pageSnapshot = await pReflect(getPageSnapshot(page))
+          const pageMeta = pageSnapshot.isRejected ? {} : pageSnapshot.value
+          const pageReadyResult = await pReflect(
+            isPageReady({
+              page,
+              screenshot,
+              isWhite,
+              isWhiteScreenshot,
+              ...pageMeta
+            })
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/pdf/src/index.js]: Fixed in fa8f709c: both the fast path and the screenshot poll pass goto `response` into `isPageReady`.

File: packages/pdf/src/index.js [hunk 3]
```diff
@@ -46,23 +49,40 @@ const getMargin = unit => {
...
+
     return captureWithNavigationRetry(
-      () => page.pdf({ ...rest, margin: getMargin(margin), printBackground, scale }),
+      () =>
+        page.pdf({
+          ...rest,
+          margin: getMargin(margin),
+          printBackground,
+          scale
+        }),
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/pdf/src/index.js:64]: Fixed in fa8f709c: `render` whitelists Puppeteer `page.pdf()` options via `toPdfOpts` so navigation/readiness keys are not spread into the CDP call.

File: packages/screenshot/src/index.js [hunk 4]
```diff
@@ -227,6 +225,22 @@ module.exports = ({ goto, ...gotoOpts }) => {
...
+          await prepareFullDocument(page, { goto, timeout: scrollTimeout })
+          screenshot = await captureWithNavigationRetry(
+            async () => {
+              await pReflect(page.evaluate(expandOverflow))
+              return page.screenshot(toScreenshotOpts(opts, { fullPage: true }))
+            },
+            { page, goto, timeout: scrollTimeout }
+          )
+          isWhite = await isWhiteScreenshot(screenshot)
+        }
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/screenshot/src/index.js:223]: Addressed in fa8f709c: fullPage/PDF readiness polls now do one bounded mid-poll `scrollFullPageToLoadContent` when the probe is non-white, so lazy/overflow shells can paint before readiness times out. Final `prepareFullDocument` remains gated on `isReady` so bot/verification pages are not expanded.

File: packages/pdf/src/index.js [hunk 5]
```diff
@@ -76,11 +96,14 @@ module.exports = ({ goto, ...gotoOpts } = {}) => {
...
       scale,
       printBackground,
+      mediaType = PDF_DEFAULT_OPTS.mediaType,
       waitUntil = PDF_DEFAULT_OPTS.waitUntil,
       waitForDom = PDF_DEFAULT_OPTS.waitForDom,
+      isPageReady = PDF_DEFAULT_OPTS.isPageReady,
       ...rest
     } = opts
     const waitForDomOpts = resolveWaitForDom(waitForDom)
+    const gotoOpts = { ...rest, mediaType }
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/pdf/src/index.js]: Intentional: `mediaType` stays unset so Chromium applies print CSS inside `page.pdf()` without an extra `emulateMediaType`. Callers that need screen (or explicit print emulation) pass `mediaType` — e.g. CLI `--mediaType=screen`. Documented on `PDF_DEFAULT_OPTS` / prepare destructure.

File: packages/pdf/src/index.js [hunk 6]
```diff
@@ -104,82 +127,82 @@ module.exports = ({ goto, ...gotoOpts } = {}) => {
...
-      // pixel), or enough visible text with webfonts loaded (a pending
-      // `font-display: block` font renders text invisible, exactly when a
-      // capture would be white) — and that content is not `covered` by an
-      // opaque viewport-filling layer (a fixed white loading overlay passes
-      // every other DOM signal while a capture stays white): skip the
-      // screenshot poll. Height and viewport come from the same in-page
-      // snapshot (`page.viewport()` is null under `defaultViewport: null`),
-      // and an unknown viewport skips the fast path rather than dropping the
-      // taller-than-viewport guard. A gate that timed out never settled, so
-      // don't trust its partial snapshot: fall through to the blank check.
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/pdf/src/index.js:143]: Fixed in fa8f709c: `prepare` merges `{ ...readiness, ...prep }` so `timedOut` and other waitForReady fields are preserved.

File: packages/pdf/src/index.js [hunk 7]
```diff
@@ -46,25 +97,42 @@ const getMargin = unit => {
...
-    )
+  const render = async (page, opts = {}) => {
+    const pdfOpts = toPdfOpts({
+      ...opts,
+      margin: opts.margin ?? PDF_DEFAULT_OPTS.margin,
+      scale: opts.scale ?? PDF_DEFAULT_OPTS.scale,
+      printBackground: opts.printBackground ?? PDF_DEFAULT_OPTS.printBackground
+    })
+
+    await pReflect(page.evaluate(expandOverflow))
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/pdf/src/index.js]: Fixed: `render` only runs `expandOverflow` when `prepare` marked the document ready (`documentPrepared`). Auto-mode pages that fail `isPageReady` are no longer unwrapped at print time.

File: packages/screenshot/src/index.js [hunk 8]
```diff
@@ -223,10 +222,35 @@ module.exports = ({ goto, ...gotoOpts }) => {
...
+
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/screenshot/src/index.js:224]: Fixed: when `fullPage` is set, the final capture is always a fullPage screenshot. Readiness probes stay viewport-only; `prepareFullDocument` / overflow unwrap remain gated on `isReady`.

File: packages/pdf/src/index.js [hunk 9]
```diff
@@ -94,92 +175,116 @@ module.exports = ({ goto, ...gotoOpts } = {}) => {
...
+    await goto(page, { ...gotoOpts, url, waitUntil, waitUntilAuto })
+
+    // Same full-document prep as fullPage screenshot (scroll + unwrap overflow).
+    if (isReady) {
+      const prep = await prepareFullDocument(page, {
+        goto,
+        timeout: resolveScrollTimeout(goto, rest.timeout)
+      })
+      readiness = { ...readiness, ...prep }
+      documentPrepared = true
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/pdf/src/index.js]: Fixed: `documentPrepared` is now a `WeakMap` keyed by `page`, so overlapping prepare/render jobs on the same factory no longer clobber each other.

File: packages/screenshot/src/prepare-full-document.js [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        let pageHeight = doc ? doc.scrollHeight : 0
+        let viewport = window.innerHeight
+
+        if (pageHeight <= viewport + 1 && document.body) {
+          let best = null
+          for (const el of document.body.querySelectorAll('*')) {
+            if (el.scrollHeight <= el.clientHeight + 20) continue
+            const { overflowY } = window.getComputedStyle(el)
+            if (overflowY !== 'auto' && overflowY !== 'scroll') continue
+            if (!best || el.scrollHeight > best.scrollHeight) best = el
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/screenshot/src/prepare-full-document.js]: Fixed: scroll / expand / overflow-wait now share `OVERFLOW_MIN_PX` (200), so hydration and unwrap pick the same scroller.

File: packages/pdf/src/index.js [hunk 11]
```diff
@@ -94,92 +136,116 @@ module.exports = ({ goto, ...gotoOpts } = {}) => {
...
+            duration: screenshotTime()
+          })
+        } while (!isReady && elapsed() < pollTimeout)
       }
 
-      // Otherwise fall back to the screenshot poll — re-wait while the first
-      // paint is still blank — to keep the blank-SPA protection. The page has
-      // already settled, so the first capture won't race a navigation, and a
-      // non-blank page exits after that single shot without an extra re-wait.
-      let isWhite = false
```
[the remark this answers is not in the document]
Comment (Kikobeats) [reply packages/pdf/src/index.js:159]: Fixed: non-`auto` PDF `prepare` now runs `prepareFullDocument` (same as fullPage screenshot with an explicit `waitUntil`) before marking the page prepared.
